homeassistant/components/wiffi/__init__.py

```diff
@@ -138,8 +145,14 @@ def _periodic_tick(self, now=None):
 class WiffiEntity(Entity):
     """Common functionality for all wiffi entities."""
 
-    def __init__(self, device, metric, options):
+    def __init__(self, device, metric, options, entity_description_type):
         """Initialize the base elements of a wiffi entity."""
+        self.entity_description = entity_description_type(
+            key=metric.description,
+            name=metric.description,
+            entity_category=NAME_TO_ENTITY_CAT.get(metric.description),
+        )
```

Comment:
Create a tuple of entity descriptions outside of this class, and pass the description into here instead.

For examples, see e.g., Forecast Solar (any many others)

Reply:
There are different Wiffi devices (weather station, pulse counter, ...). Unfortunately, the entities provided by the different devices are not documented (and I don't own all types of devices). Therefore, this integration has to guess various EntityDescription attributes from the incoming data. But this also enables support for any type of device.

Thats the reason why there is not static list of EntityDescriptions, but (in this example) a map of some well-known names to entity_categories. There are other places in the code at which other attributes are derived from the incoming data.

Reply:
I'm confused on why that matters? We should pass in EntityDescriptions and not create them the constructor of the entity.

Reply:
Ok, now I feel a little bit lost. If I got you right, you want me to create a static list of EntityDescriptions (to avoid creating them in a constructor). But I don't have the list because I don't all measurements of all devices. Thats why I create them dynamically based on the metrics provided in the incoming data.
Would it be ok to create them outside of the entity and then pass it in?

Reply:
Yes, EntityDescription are created outside the entity.
When checking which measurements the device provides, check find the matching entity description and pass that into the entity.

Or, the other common approach, iterate over all available possible entity description, check if the devices provides it, if so, pass the description into a new entity.